File: cmd/cpages/device_login.go
package main

import (
	"bytes"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

type discoveryResponse struct {
	ControlURL          string `json:"control_url"`
	ContentURL          string `json:"content_url"`
	DeviceAuthorization bool   `json:"device_authorization"`
}

type deviceCodeResponse struct {
	DeviceCode              string `json:"device_code"`
	UserCode                string `json:"user_code"`
	VerificationURI         string `json:"verification_uri"`
	VerificationURIComplete string `json:"verification_uri_complete"`
	ExpiresIn               int    `json:"expires_in"`
	Interval                int    `json:"interval"`
}

type discoveryHTTPError struct {
	Status int
}

func (e *discoveryHTTPError) Error() string {
	return fmt.Sprintf("server returned HTTP %d", e.Status)
}

func cmdLogin(args []string) error {
	fs := flag.NewFlagSet("login", flag.ContinueOnError)
	serverFlag := fs.String("server", "", "content or control server URL (prompted if omitted)")
	legacy := fs.Bool("legacy-passcode", false, "prompt for the legacy shared passcode")
	force := fs.Bool("force", false, "save a legacy login even when it cannot be verified")
	deviceName := fs.String("device-name", "", "label shown to the owner during approval")
	readOnly := fs.Bool("read-only", false, "request only pages:read")
	if err := parse(fs, args); err != nil {
		return err
	}
	if fs.NArg() != 0 {
		return errors.New("login does not accept positional arguments")
	}
	if *legacy {
		if *deviceName != "" || *readOnly {
			return errors.New("--device-name and --read-only cannot be used with --legacy-passcode")
		}
		return cmdLegacyLogin(*serverFlag, *force)
	}
	if *force {
		return errors.New("--force is available only with --legacy-passcode")
	}
	return loginWithDevice(*serverFlag, *deviceName, *readOnly)
}

func loginWithDevice(serverValue, deviceName string, readOnly bool) error {
	existing, err := loadConfig()
	if err != nil {
		return err
	}
	server := firstNonEmpty(serverValue, os.Getenv("COLUMBIA_PAGES_URL"))
	if server == "" {
		hint := ""
		if existing.URL != "" {
			hint = " [" + existing.URL + "]"
		}
		value, err := readLine("Server URL" + hint + ": ")
		if err != nil {
			return err
		}
		server = firstNonEmpty(value, existing.URL)
	}
	server, err = normalizeServerURL(server)
	if err != nil {
		return err
	}

	discovery, err := discover(server)
	if err != nil {
		var httpErr *discoveryHTTPError
		if errors.As(err, &httpErr) && httpErr.Status == http.StatusNotFound {
			return fmt.Errorf("device discovery failed: %w; if this is an older instance and the owner enabled legacy access, retry with `cpages login --legacy-passcode --server %s`", err, server)
		}
		return fmt.Errorf("device discovery failed: %w", err)
	}
	if !discovery.DeviceAuthorization || discovery.ControlURL == "" {
		return errors.New("this instance does not support device login; use --legacy-passcode only if the owner has enabled migration access")
	}
	controlURL, err := normalizeServerURL(discovery.ControlURL)
	if err != nil {
		return fmt.Errorf("server returned an unsafe control URL: %w", err)
	}
	if deviceName == "" {
		host, hostErr := os.Hostname()
		if hostErr != nil || strings.TrimSpace(host) == "" {
			host = "this device"
		}
		deviceName = "cpages on " + host
	}
	if len(deviceName) > 120 {
		return errors.New("--device-name must be 120 characters or fewer")
	}
	secret, err := randomDeviceSecret()
	if err != nil {
		return err
	}
	scopes := []string{"pages:read", "pages:write"}
	if readOnly {
		scopes = []string{"pages:read"}
	}
	code, err := requestDeviceCode(controlURL, secret, deviceName, scopes)
	if err != nil {
		return err
	}
	fmt.Printf("Open this URL to approve the device:\n%s\n\nCode: %s\n\nWaiting for approval...\n", code.VerificationURIComplete, code.UserCode)
	token, err := pollDeviceToken(controlURL, code, secret)
	if err != nil {
		return err
	}
	if err := saveConfig(config{URL: controlURL, Token: token}); err != nil {
		return err
	}
	fmt.Printf("✓ Logged in to %s\n", controlURL)
	fmt.Printf("  saved to %s\n", configPath())
	return nil
}

func discover(server string) (discoveryResponse, error) {
	resp, err := (&http.Client{Timeout: 15 * time.Second}).Get(server + "/.well-known/columbia-pages")
	if err != nil {
		return discoveryResponse{}, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return discoveryResponse{}, &discoveryHTTPError{Status: resp.StatusCode}
	}
	var discovery discoveryResponse
	if err := json.NewDecoder(resp.Body).Decode(&discovery); err != nil {
		return discovery, err
	}
	return discovery, nil
}

func requestDeviceCode(controlURL, secret, label string, scopes []string) (deviceCodeResponse, error) {
	payload, err := json.Marshal(map[string]any{"device_secret": secret, "device_label": label, "scopes": scopes})
	if err != nil {
		return deviceCodeResponse{}, err
	}
	resp, err := (&http.Client{Timeout: 15 * time.Second}).Post(controlURL+"/api/auth/device/code", "application/json", bytes.NewReader(payload))
	if err != nil {
		return deviceCodeResponse{}, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusCreated {
		return deviceCodeResponse{}, responseError(resp)
	}
	var code deviceCodeResponse
	if err := json.NewDecoder(resp.Body).Decode(&code); err != nil {
		return code, err
	}
	if code.DeviceCode == "" || code.UserCode == "" || code.VerificationURI == "" || code.ExpiresIn <= 0 {
		return code, errors.New("server returned an incomplete device authorization response")
	}
	if code.Interval < 1 {
		code.Interval = 5
	}
	return code, nil
}

func pollDeviceToken(controlURL string, code deviceCodeResponse, secret string) (string, error) {
	deadline := time.Now().Add(time.Duration(code.ExpiresIn) * time.Second)
	interval := time.Duration(code.Interval) * time.Second
	for time.Now().Before(deadline) {
		payload, err := json.Marshal(map[string]string{"device_code": code.DeviceCode, "device_secret": secret})
		if err != nil {
			return "", err
		}
		resp, err := (&http.Client{Timeout: 15 * time.Second}).Post(controlURL+"/api/auth/device/token", "application/json", bytes.NewReader(payload))
		if err != nil {
			return "", err
		}
		var result struct {
			AccessToken string `json:"access_token"`
			Error       string `json:"error"`
		}
		decodeErr := json.NewDecoder(resp.Body).Decode(&result)
		resp.Body.Close()
		if decodeErr != nil {
			return "", decodeErr
		}
		if resp.StatusCode == http.StatusOK && result.AccessToken != "" {
			return result.AccessToken, nil
		}
		switch result.Error {
		case "authorization_pending":
		case "slow_down":
			if retry, err := time.ParseDuration(resp.Header.Get("Retry-After") + "s"); err == nil && retry > interval {
				interval = retry
			} else {
				interval += 5 * time.Second
			}
		case "access_denied":
			return "", errors.New("device authorization was denied")
		case "expired_token":
			return "", errors.New("device authorization expired")
		default:
			return "", fmt.Errorf("device authorization failed: %s", result.Error)
		}
		time.Sleep(interval)
	}
	return "", errors.New("device authorization expired")
}

func randomDeviceSecret() (string, error) {
	value := make([]byte, 32)
	if _, err := rand.Read(value); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(value), nil
}

func responseError(resp *http.Response) error {
	data, _ := io.ReadAll(resp.Body)
	var body struct {
		Error string `json:"error"`
	}
	if json.Unmarshal(data, &body) == nil && body.Error != "" {
		return fmt.Errorf("server %d: %s", resp.StatusCode, body.Error)
	}
	return fmt.Errorf("server %d: %s", resp.StatusCode, strings.TrimSpace(string(data)))
}

File: ts/src/cli/deviceLogin.ts
import { hostname } from "node:os"
import { Clock, Config, Console, Duration, Effect, Option, Schema } from "effect"
import { Prompt } from "effect/unstable/cli"
import { HttpClient, HttpClientRequest } from "effect/unstable/http"
import {
  DeviceCodeResponse,
  DeviceTokenResponse,
  DiscoveryResponse,
  SCOPE_READ,
  SCOPE_WRITE
} from "../shared/api.ts"
import { randomBase64 } from "../shared/crypto.ts"
import { normalizeOrFail } from "./client.ts"
import { CliConfig } from "./config.ts"
import { CliUserError, cliError, serverErrorMessage } from "./errors.ts"

const messageOf = (error: unknown): string =>
  error instanceof CliUserError ? error.message : String(error)

const discover = (http: HttpClient.HttpClient, server: string) =>
  Effect.gen(function* () {
    const response = yield* http.execute(
      HttpClientRequest.get(`${server}/.well-known/columbia-pages`)
    )
    if (response.status !== 200) {
      return yield* cliError(`server returned HTTP ${response.status}`)
    }
    const json = yield* response.json
    return yield* Schema.decodeUnknownEffect(DiscoveryResponse)(json)
  }).pipe(Effect.timeout(Duration.seconds(15)))

const requestDeviceCode = (
  http: HttpClient.HttpClient,
  controlUrl: string,
  secret: string,
  label: string,
  scopes: ReadonlyArray<string>
) =>
  Effect.gen(function* () {
    const response = yield* http.execute(
      HttpClientRequest.post(`${controlUrl}/api/auth/device/code`).pipe(
        HttpClientRequest.bodyJsonUnsafe({
          device_secret: secret,
          device_label: label,
          scopes
        })
      )
    )
    const text = yield* response.text
    if (response.status !== 201) {
      return yield* cliError(serverErrorMessage(response.status, text))
    }
    const parsed = yield* Effect.try({
      try: () => JSON.parse(text) as unknown,
      catch: (e) => cliError(messageOf(e))
    })
    const code = yield* Schema.decodeUnknownEffect(DeviceCodeResponse)(parsed)
    if (
      code.device_code === "" ||
      code.user_code === "" ||
      code.verification_uri === "" ||
      code.expires_in <= 0
    ) {
      return yield* cliError("server returned an incomplete device authorization response")
    }
    return code.interval < 1 ? { ...code, interval: 5 } : code
  }).pipe(
    Effect.timeout(Duration.seconds(15)),
    Effect.mapError((e) => (e instanceof CliUserError ? e : cliError(messageOf(e))))
  )

const pollDeviceToken = (
  http: HttpClient.HttpClient,
  controlUrl: string,
  code: typeof DeviceCodeResponse.Type,
  secret: string
) =>
  Effect.gen(function* () {
    const start = yield* Clock.currentTimeMillis
    const deadline = start + code.expires_in * 1000
    let interval = code.interval
    while ((yield* Clock.currentTimeMillis) < deadline) {
      const poll = yield* Effect.gen(function* () {
        const response = yield* http.execute(
          HttpClientRequest.post(`${controlUrl}/api/auth/device/token`).pipe(
            HttpClientRequest.bodyJsonUnsafe({
              device_code: code.device_code,
              device_secret: secret
            })
          )
        )
        const json = yield* response.json
        const result = yield* Schema.decodeUnknownEffect(DeviceTokenResponse)(json)
        return { response, result }
      }).pipe(
        Effect.timeout(Duration.seconds(15)),
        Effect.mapError((e) => (e instanceof CliUserError ? e : cliError(messageOf(e))))
      )
      const { response, result } = poll
      const accessToken = result.access_token ?? ""
      if (response.status === 200 && accessToken !== "") return accessToken
      switch (result.error ?? "") {
        case "authorization_pending":
          break
        case "slow_down": {
          const retryAfter = Number(response.headers["retry-after"] ?? "")
          if (Number.isFinite(retryAfter) && retryAfter > interval) interval = retryAfter
          else interval += 5
          break
        }
        case "access_denied":
          return yield* cliError("device authorization was denied")
        case "expired_token":
          return yield* cliError("device authorization expired")
        default:
          return yield* cliError(`device authorization failed: ${result.error ?? ""}`)
      }
      yield* Effect.sleep(Duration.seconds(interval))
    }
    return yield* cliError("device authorization expired")
  })

export const loginWithDevice = (
  serverFlag: Option.Option<string>,
  deviceNameFlag: Option.Option<string>,
  readOnly: boolean
) =>
  Effect.gen(function* () {
    const cliConfig = yield* CliConfig
    const http = yield* HttpClient.HttpClient
    const existing = yield* cliConfig.load

    const envUrl = yield* Config.string("COLUMBIA_PAGES_URL").pipe(
      Config.withDefault(""),
      Effect.mapError((e) => cliError(String(e)))
    )
    let server = Option.getOrElse(serverFlag, () => "").trim()
    if (server === "") server = envUrl.trim()
    if (server === "") {
      const hint = existing.url !== "" ? ` [${existing.url}]` : ""
      const entered = yield* Prompt.text({
        message: `Server URL${hint}:`,
        ...(existing.url !== "" ? { default: existing.url } : {})
      }).pipe(Effect.mapError(() => cliError("login cancelled")))
      server = entered.trim() !== "" ? entered : existing.url
    }
    const normalizedServer = yield* normalizeOrFail(server)

    const discovery = yield* discover(http, normalizedServer).pipe(
      Effect.mapError((e) =>
        cliError(
          `device discovery failed: ${messageOf(e)}; upgrade the Columbia Pages deployment, then retry`
        )
      )
    )
    if (!discovery.device_authorization || discovery.control_url === "") {
      return yield* cliError(
          "this instance does not support device login; upgrade the Columbia Pages deployment, then retry"
        )
    }
    const controlUrl = yield* normalizeOrFail(discovery.control_url).pipe(
      Effect.mapError((e) => cliError(`server returned an unsafe control URL: ${e.message}`))
    )

    let deviceName = Option.getOrElse(deviceNameFlag, () => "")
    if (deviceName === "") {
      const host = hostname().trim()
      deviceName = `cpages on ${host !== "" ? host : "this device"}`
    }
    if (deviceName.length > 120) {
      return yield* cliError("--device-name must be 120 characters or fewer")
    }

    const secret = yield* randomBase64(32)
    const scopes = readOnly ? [SCOPE_READ] : [SCOPE_READ, SCOPE_WRITE]
    const code = yield* requestDeviceCode(http, controlUrl, secret, deviceName, scopes)

    yield* Console.log(
      `Open this URL to approve the device:\n${code.verification_uri_complete}\n\nCode: ${code.user_code}\n\nWaiting for approval...`
    )
    const token = yield* pollDeviceToken(http, controlUrl, code, secret)

    yield* cliConfig.save({ url: controlUrl, token })
    yield* Console.log(`✓ Logged in to ${controlUrl}`)
    yield* Console.log(`  saved to ${cliConfig.path}`)
  })
